### PR TITLE
Add ROX storage tests

### LINKED ISSUE #1088: read only volumes should not be expanded and we need OSS ROX testing
#972 added unconditional filesystem resize and it should not happen for read only volumes. We are not seeing failures in mounting as resizing RO volume only produce an warning but this could have broken ROX functionality.

1. we should not try to expand RO volumes
2. we should write an OSS ROX test ASAP

### PR DESCRIPTION
**Pull request by judemars**
#### What this PR does / why we need it:

Adds ROX k8s CSI integration tests. 

Adds tests for restoring from snapshot and clone in ReadOnlyMany mode. It tests that the new pvcs contain the source data, as well as that the new volumes are not resized (since the fs is readonly).

Some of the resize testing code is based from https://github.com/kubernetes/kubernetes/pull/110304/files (adding test for fs resize) which isn't merged yet at the time of this writing. 

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/1088

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/e2e/storage/testsuites/provisioning.go [hunk 1]
```diff
@@ -440,6 +443,231 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
 		l.testCase.TestDynamicProvisioning(ctx)
 	})
 
+	ginkgo.It("should provision storage with pvc data source for readonly filesystem", func(ctx context.Context) {
```
Comment (leiyiz) [inline test/e2e/storage/testsuites/provisioning.go:446]: I wonder if there should be a [Feature:xxxx] similar to `[Feature:VolumeSnapshotDataSource]` above. This kind of tags have been used for test skipping and/or focusing purposes.

Comment (mattcary) [reply test/e2e/storage/testsuites/provisioning.go:446]: Feature: tags basically mean the tests are off by default. I think we want these tests on by default?

Comment (leiyiz) [reply test/e2e/storage/testsuites/provisioning.go:446]: ahh right. make sense

File: test/e2e/storage/testsuites/provisioning.go [hunk 2]
```diff
@@ -1139,6 +1371,55 @@ func preparePVCDataSourceForProvisioning(
 	return dataSourceRef
 }
 
+// findVolumeMountPath looks for a claim name inside a pod and returns an absolute path of its volume mount point.
```
Comment (RomanBednar) [inline test/e2e/storage/testsuites/provisioning.go:1374]: We probably don't need to copy this code from open [PR](https://github.com/kubernetes/kubernetes/pull/110304). I believe it could be merged soon as it was only waiting for getting fixed version of the driver into [CI](https://github.com/kubernetes/kubernetes/pull/114306) which is now passing. @mattcary what do you think?

**Comment by RomanBednar** [comment]
@judemars thanks for the fixes for ROX case! The changes look good to me, I'd only suggest cleaning up those commits a bit. Maybe squash the ones resolving conflicts and have the refactoring in one commit and the actual test addition in a separate one. 

Also if we decide to wait for the merges I [mentioned](https://github.com/kubernetes/kubernetes/pull/114628#discussion_r1059936882) or until the commits are cleaned up this PR should be marked as WIP.

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:1374]: This PR depends on findVolumeMountPath existing as both ROX tests use it. What is the ETA on that PR merging? I'd be happy to wait for it to merge in if it's within this month (Jan 2023), but beyond that I think it needs to be merged in so we can get this use case protected.

**Comment by judemars** [comment]
@RomanBednar I added the /label tide/merge-method-squash label so it will be squashed when the PR merges. LMK if you feel strongly that they should be two separate changes though and I'll break it up, I just generally find the auto-squash easier to work with.

**Comment by bart0sh** [comment]
@judemars please, rebase your PR, squish commits and get rid of ```do-not-merge/contains-merge-commits``` label

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (RomanBednar) [reply test/e2e/storage/testsuites/provisioning.go:1374]: Agreed, unfortunately the ETA is not fully up to me. @mattcary do you see the PRs I mentioned getting merged any time soon?

Comment (mattcary) [reply test/e2e/storage/testsuites/provisioning.go:1374]: Ah, sorry, I missed the comment  on #110304.

I think it will be most expedient on that PR to just update the hostpath driver, then that PR can be submitted. I can then make a PR that updates the PD CSI version that the e2e tests use & change  InitGcePDCSIDriver.

That should unblock this PR I think as there's no GCE PD dependency?

**Comment by judemars** [comment]
Thanks @bart0sh, fixed now!

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (mattcary) [reply test/e2e/storage/testsuites/provisioning.go:1374]: Yes, please, go with #2. Thanks!

Comment (mattcary) [reply test/e2e/storage/testsuites/provisioning.go:1374]: (I just created #114955 which goes into more detail of why I think #2 is correct :-)

Comment (RomanBednar) [reply test/e2e/storage/testsuites/provisioning.go:1374]: @mattcary Done. I had to also add the recently added [anti-capability](https://github.com/kubernetes/kubernetes/pull/110304/commits/c78bb5a43b5e4465bfdc2d6fe56f318743207c34#diff-b3ae4c017c06f067fbe5d45cd922c508eccfa25195011970a9426a35d2115cc7R826) to pass the CI.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:1374]: Looks like go/k8spr/110304 needs another lgtm after some changes/squashing - @mattcary do you have time to look? We're waiting on that to go in before this one.

**Comment by SergeyKanzhelev** [comment]
/cc 

for sig node changes

File: test/e2e/storage/testsuites/provisioning.go (hunk 2 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:1374]: Gentle ping @mattcary

File: test/e2e/storage/testsuites/provisioning.go [hunk 3]
```diff
@@ -440,6 +443,231 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
 
+	ginkgo.It("should provision storage with pvc data source for readonly filesystem", func(ctx context.Context) {
+		if !dInfo.Capabilities[storageframework.CapPVCDataSource] {
+			e2eskipper.Skipf("Driver %q does not support cloning - skipping", dInfo.Name)
+		}
+
+		init(ctx)
+
+		if l.config.ClientNodeSelection.Name == "" {
+			// Schedule all pods to the same topology segment (e.g. a cloud availability zone), some
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:454]: I believe https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/1150 should fix the topology issue with cloning.

File: test/e2e/storage/testsuites/provisioning.go [hunk 4]
```diff
@@ -440,6 +443,231 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+		// Create claim w/ bigger size than original
+		biggerROClaim := l.testCase.Claim.DeepCopy()
+		biggerROClaim.Name = "bigger-claim"
+		biggerROClaim.Spec.Resources = v1.ResourceRequirements{
+			Requests: v1.ResourceList{
+				v1.ResourceStorage: biggerSize,
+			},
+		}
+		biggerROClaim.Spec.AccessModes = []v1.PersistentVolumeAccessMode{
+			v1.PersistentVolumeAccessMode(v1.ReadOnlyMany),
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:501]: Not all volume types support ROX access mode. You may need to guard these test cases with a new test capability.

File: test/e2e/storage/testsuites/provisioning.go [hunk 5]
```diff
@@ -440,6 +443,231 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+		e2evolume.WaitForVolumeAttachmentTerminated(ctx, volumeAttachment, f.ClientSet, f.Timeouts.DataSourceProvision)
+
+		originalClaim, err := l.testCase.Client.CoreV1().PersistentVolumeClaims(l.pvc.Namespace).Get(ctx, l.sourcePVC.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "Failed to get pvc: %v", err)
+
+		var originFSSize int
+		if pattern.VolMode != "Block" {
+			// Create pod to get mount path of fs
+			createdClaims := []*v1.PersistentVolumeClaim{originalClaim}
+			pod, err := e2epod.CreatePod(ctx, l.testCase.Client, f.Namespace.Name, nil, createdClaims, true, "", true)
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:479]: Have you looked into using `CreateSecPod`? https://github.com/kubernetes/kubernetes/blob/0a25370cd896bf15a7301be93e8ab347974b7da4/test/e2e/framework/pod/create.go#L105

It puts all the pod config in a struct so you could potentially avoid updating a bunch of unrelated test cases.

File: test/e2e/storage/testsuites/provisioning.go (hunk 3 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:539]: Good to know. I'll file an internal bug to clean this up after this goes in, as I don't think we should block and wait for another PR to go in (the last one took months)

File: test/e2e/storage/testsuites/provisioning.go (hunk 4 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:586]: Added CapReadOnlyMany, good idea thanks! Do I need to go and opt some drivers out or is the onus on driver owners to do this if their k8s integration tests start to fail?

File: test/e2e/storage/testsuites/provisioning.go (hunk 5 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go]: This is great, thanks! Updated.

File: test/e2e/storage/testsuites/provisioning.go [hunk 6]
```diff
@@ -527,6 +528,266 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+
+			// Get actual size of the restored filesystem.
+			restoredFSSize, err := getFilesystemSizeBytes(pod, mountpath)
+			framework.ExpectNoError(err, "Failed to obtain filesystem size of a volume mount: %v", err)
+
+			msg := fmt.Sprintf("Filesystem resize occurred even though it is readonly. "+
+				"Restored fs size: %v bytes is not the same as origin fs size: %v bytes.\n"+
+				"HINT: Your driver needs to check the volume and skip resize if it's readonly.",
+				restoredFSSize, originFSSize)
+			gomega.Expect(restoredFSSize).Should(gomega.BeEquivalentTo(originFSSize), msg)
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:787]: I am not sure if we can require equivalence in capacity even if it's ROX. I think NFS-like volume types can support a larger capacity despite being read-only. I think the proper behavior for plugins that can't support the larger size is to fail the CreateVolume call.

@saikat-royc @gnufied  wdyt?

File: test/e2e/storage/testsuites/provisioning.go [hunk 7]
```diff
@@ -527,6 +528,266 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+		testConfig := storageframework.ConvertTestConfig(l.config)
+		expectedContent := fmt.Sprintf("Hello from namespace %s", f.Namespace.Name)
+
+		l.sourcePVC.Name = "source-pvc"
+		dataSourceRef := preparePVCDataSourceForProvisioning(ctx, f, testConfig, l.cs, l.sourcePVC, l.sc, pattern.VolMode, expectedContent)
+		l.pvc.Spec.DataSourceRef = dataSourceRef
+		l.testCase.NodeSelection = testConfig.ClientNodeSelection
+
+		// Cloning fails if the source disk is still in the process of detaching, so we wait for the VolumeAttachment to be removed before cloning.
+		volumeAttachment := e2evolume.GetVolumeAttachmentName(ctx, f.ClientSet, testConfig, l.testCase.Provisioner, dataSourceRef.Name, l.sourcePVC.Namespace)
```
Comment (gnufied) [inline test/e2e/storage/testsuites/provisioning.go:558]: Are all ROX volume types going to be attachable? Does the code fails gracefully when it is not attachable?

File: test/e2e/storage/testsuites/provisioning.go (hunk 6 above)
Comment (gnufied) [reply test/e2e/storage/testsuites/provisioning.go:787]: Yep. This seems like a problem in case nfs volumes are being provisioned off same directory or something.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:787]: Thanks for pointing this out. The main goal here was just to make sure that a resize didn't occur when it shouldn't - maybe we add a anti-capability here and only opt-in drivers that we believe should have the volume be the same size before and after (PD, filestore, etc?)

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:558]: We have a capability for this in storageframework.CapReadOnlyMany if that's what you mean? So this test is only scoped to PD and hostpath driver which do support ROX

File: test/e2e/storage/testsuites/provisioning.go (hunk 4 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:596]: Ah it looks like Capabilities are opt-in so opting in PD in this PR is enough I think? lmk if I'm missing something though

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go:558]: The storageframework is used by many other plugins outside of k/k. So from a general standpoint, we cannot assume that drivers that do support ROX also support attachments.

I think in this case though, the volumeattachment object would not exist so it would pass?

File: test/e2e/storage/testsuites/provisioning.go (hunk 6 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go:787]: I think we can leverage the resizeFromRestore anti capability: https://github.com/kubernetes/kubernetes/pull/114628/files#diff-d7be57591f195129598536ab483c8e93bc048d836c09ed47b409e6e6b5579bffR189

In the case of PD driver, it can't support resize from restore for ROX.

File: test/e2e/storage/testsuites/provisioning.go (hunk 7 above)
Comment (gnufied) [reply test/e2e/storage/testsuites/provisioning.go:558]: Yeah it does seem okay at least for now. Ideally, it should be behind a capability.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:558]: Yup, other tests in this file also test volume attachments in the same fashion we are doing here. If we want to put all of those tests behind a shared capability I think we should do that in another PR.

File: test/e2e/storage/testsuites/provisioning.go (hunk 6 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:787]: PD driver does support general resize from restore as of https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/972 and that's what that as per https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go#L829 that anti-capability will soon be removed from PD. 

I think we need an anti-capability here that's specific to Resize+ROX. I'll add CapFSROXResizeFromSourceNotSupported.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:792]: Alrighty added CapFSROXResizeFromSourceNotSupported to this test, but down below right before we check filesystem size. Before that, this test also checks that the clone occurred properly, which I think nfs, etc would still benefit from.

File: test/e2e/storage/testsuites/provisioning.go [hunk 8]
```diff
@@ -527,6 +528,271 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+
+			// Save filesystem size of the restored volume.
+			restoredFSSize, err := getFilesystemSizeBytes(pod, mountpath)
+			framework.ExpectNoError(err, "Failed to obtain filesystem size of a volume mount: %v", err)
+
+			msg := fmt.Sprintf("Filesystem resize occurred even though it is readonly. "+
+				"Restored fs size: %v bytes is not the same as origin fs size: %v bytes.\n"+
+				"HINT: Your driver needs to check the volume and skip resize if it's readonly.",
+				restoredFSSize, originFSSize)
+			gomega.Expect(restoredFSSize).Should(gomega.BeEquivalentTo(originFSSize), msg)
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:659]: In the case where the sameSize claim is created, the restored size should be the same.

In the case where the biggerSize claim is created, the restored size should be bigger.

File: test/e2e/storage/testsuites/provisioning.go [hunk 9]
```diff
@@ -527,6 +528,271 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
...
+					Volume:          *storageutils.CreateVolumeSource(claim.Name, false /* readOnly */),
+					Mode:            pattern.VolMode,
+					File:            "index.html",
+					ExpectedContent: expectedContent,
+				},
+			}
+			e2evolume.TestVolumeClientSlow(ctx, f, testConfig, nil, "", tests)
+		}
+
+		for _, claim := range []*v1.PersistentVolumeClaim{sameSizeROClaim, biggerROClaim} {
```
Comment (msau42) [inline test/e2e/storage/testsuites/provisioning.go:621]: I think this should be two different test cases (`It()`), rather than two cases under one.

Also I think from a strictly "conformant" angle, drivers that can't support expand for rox, should return an error if someone tries to create a PVC with a bigger size, so I think the bigger PVC test case should not run at all in that case.

File: test/e2e/storage/testsuites/provisioning.go (hunk 8 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:659]: The whole point of this test is to test that in ROX mode, no matter what size you request in the PVC, the cloned size will be the same as the original, since a resize cannot occur (bc the fs is readonly). Are we not aligned on this? See https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/issues/1088 and https://github.com/kubernetes/kubernetes/pull/114325 for context. I also have an internal test plan I'll send along.

It's a little hard to follow because we have both l.testCase.Claim and l.pvc, but you can see [here](https://github.com/kubernetes/kubernetes/blob/dd985dcdcde8d9f2b89caec382fdf9db29b845dc/test/e2e/storage/testsuites/provisioning.go#L172) they are the same underlying object. So then, by setting:

```
l.pvc.Spec.DataSourceRef = dataSourceRef
```
above, then we create both sameSizeROClaim and biggerROClaim as deep copies of that claim:

```
biggerROClaim := l.testCase.Claim.DeepCopy()
...
sameSizeROClaim := l.testCase.Claim.DeepCopy()
```
So both of those are being restored from the same cloned data source, and even though we request a larger size in the biggerPVC, since we're in ROX mode, it should restore to the original size since it can't do a resize.

File: test/e2e/storage/testsuites/provisioning.go (hunk 9 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go:621]: Are there some drivers that *can* resize while ROX? It was my understanding it was impossible since the fs is readonly. 

Re:conformance, is there a spec somewhere that dictates we should error? This is not how the PD driver [currently works](https://docs.google.com/document/d/1vYGLfKYLnfNAE1HImQ0Gp2paY2fNKexRyfXJA1jJ0tM/edit?usp=sharing) we also need to make that change and not test for size here now.

It feels like we should just remove the size comparison code here since from the comments here it now seems to be testing undefined behavior and our main motivation here was just to add *some* test for the ROX restore behavior and that's now been delayed by many months bc of confusion around sizing. I'll do that now.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go]: Removed file size tests

File: test/e2e/storage/testsuites/provisioning.go (hunk 8 above)
Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go]: As a result of other comment I removed file size checking here so this point is now moot, so resolving.

File: test/e2e/storage/testsuites/provisioning.go (hunk 9 above)
Comment (msau42) [reply test/e2e/storage/testsuites/provisioning.go]: From the CSI spec changes in https://github.com/container-storage-interface/spec/pull/452/files, it says that if the request cannot be satisfied, then the plugin should return an error.

I think it may be too late to change the PD driver to have that behavior, because it will break workflows that used to work, so what I'm proposing here is to separate it into two tests and skip the expansion case for the PD driver.

Comment (judemars) [reply test/e2e/storage/testsuites/provisioning.go]: Makes sense, but our main focus for these tests were to make sure ROX PD so I think if we want to add that it should be in a follow up/separate effort.
